[jit] add resolveType to Resolver

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #19239 [wip][jit] use fully qualified name for ScriptClasses
* #19238 [jit] Allow for segmented printing in PythonPrint
* **#19237 [jit] add resolveType to Resolver**
* #19236 [jit] Turn resolver into a class

Now that we have first-class types, various things will need to resolve
name strings into actual typeptrs. Currently we do this just be looking
up the name in our global registry, but once we add namespacing that will
no longer be sufficient.

This PR adds the infrastructure for resolving types, although the
functionality is the same as before.

Test Plan:
jit tests

gh-metadata: pytorch pytorch 19237 gh/suo/25/head

Differential Revision: [D14928289](https://our.internmc.facebook.com/intern/diff/D14928289)

suo committed:
[jit] add resolveType to Resolver

Now that we have first-class types, various things will need to resolve
name strings into actual typeptrs. Currently we do this just be looking
up the name in our global registry, but once we add namespacing that will
no longer be sufficient.

This PR adds the infrastructure for resolving types, although the
functionality is the same as before.

Test Plan:
jit tests

suo committed:
Update on "[jit] add resolveType to Resolver"

[jit] add resolveType to Resolver

Now that we have first-class types, various things will need to resolve
name strings into actual typeptrs. Currently we do this just be looking
up the name in our global registry, but once we add namespacing that will
no longer be sufficient.

This PR adds the infrastructure for resolving types, although the
functionality is the same as before.

Test Plan:
jit tests

gh-metadata: pytorch pytorch 19237 gh/suo/25/head

suo committed:
Update on "[jit] add resolveType to Resolver"

[jit] add resolveType to Resolver

Now that we have first-class types, various things will need to resolve
name strings into actual typeptrs. Currently we do this just be looking
up the name in our global registry, but once we add namespacing that will
no longer be sufficient.

This PR adds the infrastructure for resolving types, although the
functionality is the same as before.

Test Plan:
jit tests

gh-metadata: pytorch pytorch 19237 gh/suo/25/head

suo committed:
Update on "[jit] add resolveType to Resolver"

[jit] add resolveType to Resolver

Now that we have first-class types, various things will need to resolve
name strings into actual typeptrs. Currently we do this just be looking
up the name in our global registry, but once we add namespacing that will
no longer be sufficient.

This PR adds the infrastructure for resolving types, although the
functionality is the same as before.

Test Plan:
jit tests

gh-metadata: pytorch pytorch 19237 gh/suo/25/head

suo committed:
Update on "[jit] add resolveType to Resolver"

[jit] add resolveType to Resolver

Now that we have first-class types, various things will need to resolve
name strings into actual typeptrs. Currently we do this just be looking
up the name in our global registry, but once we add namespacing that will
no longer be sufficient.

This PR adds the infrastructure for resolving types, although the
functionality is the same as before.

Test Plan:
jit tests

gh-metadata: pytorch pytorch 19237 gh/suo/25/head

driazati commented on torch/csrc/jit/script/resolver.h:
```diff
@@ -29,6 +29,9 @@ struct Resolver {
       const SourceRange& loc) const = 0;
+
+  // Resolve `name` to a TypePtr.
+  virtual TypePtr resolveType(const std::string& name) const = 0;
```
Once there are qualified names we need to be careful of `name` vs `qualname` vs `basename` to keep things clear

zdevito replied:
My vote is to use the QualName class when that change is made. Separate strings are easy to lose track of.

driazati commented on torch/csrc/jit/script/script_type_parser.h:
```diff
@@ -28,6 +33,7 @@ class TORCH_API ScriptTypeParser {
   at::TypePtr subscriptToType(
       const std::string& typeName,
       const Subscript& subscript) const;
+  ResolverPtr resolver_ = nullptr;
```
Why would this ever need to be null?

suo replied:
If we ever wanted to use a "bare" ScriptTypeParser without an outer environment to look up type names in. Doesn't seem super likely, but there's no reason why we should require a resolver.

zdevito replied:
not used yet?

suo commented:
@pytorchbot retest this please

zdevito reviewed (approved):
Looks good! But it misses the addition of the resolveType override in the proxy object added in the last commit, which is problematic because it means that will be pretty likely to happen in the future as well.

suo commented:
> Looks good! But it misses the addition of the resolveType override in the proxy object added in the last commit, which is problematic because it means that will be pretty likely to happen in the future as well.

There was indeed a small resolveType override [added](https://github.com/pytorch/pytorch/pull/19237/files#diff-c0a25c14600294e812cfbdd9ac3fcae2R2793) to MethodResolver. I think as long as we keep `Resolver` an interface-only thing, the proxy type isn't so bad as it will be forced to provide an implementation of whatever new method is added to Resolver. I will add a comment to that effect in Resolver

suo committed:
Update on "[jit] add resolveType to Resolver"

[jit] add resolveType to Resolver

Now that we have first-class types, various things will need to resolve
name strings into actual typeptrs. Currently we do this just be looking
up the name in our global registry, but once we add namespacing that will
no longer be sufficient.

This PR adds the infrastructure for resolving types, although the
functionality is the same as before.

Test Plan:
jit tests

gh-metadata: pytorch pytorch 19237 gh/suo/25/head
